programs/zstdcli.c

```diff
@@ -1163,10 +1175,23 @@ int main(int argCount, const char* argv[])
         if (adaptMin > cLevel) cLevel = adaptMin;
         if (adaptMax < cLevel) cLevel = adaptMax;
 
+        if (outDirName) {
+            if (UTIL_isDirectory(outDirName)) {
+                DISPLAY("Output of files will be in directory: %s\n", outDirName);
+                dstFilenameTable = (char**)malloc(filenameIdx * sizeof(char*));
```

Comment:
Always check is a `malloc()` was successful.

Reply:
Also, what's the purpose of `dstFilenameTable` ?

Reply:
We use `dstFilenameTable` to hold the output-dir + filename string, such that the file from `filenameTable[i]` will end up being (de)compressed into `dstFilenameTable[i]`.

In `FIO_compressMultipleFilenames()', the usual behavior is to just use `filenameTable` and append or remove '.zst' to get our final path to the file, so in order to output files to a new directory, we need to create a new mapping from original path to new path based on the specified output dir.

Reply:
My understanding is as follows :
given a list of files : `{ file1 file2 file3 }`
the algorithm starts by creating a new list (table) of files, prefixing the destination directory in front of each one of them : `{ outdir/file1 outdir/file2 outdir/file3 }`.
This new list is passed to `FIO_compressMultipleFilenames()` which, as usual, generates the final name on the fly : `outdir/file1.zst` .

A question is :
why not generating the directly final name completely on the fly ?
`(outdir, file1) => outdir/file1.zst` .

A possible answer : to reduce modifications inside `FIO_compressMultipleFilenames()`.

But `FIO_compressMultipleFilenames ()` is not exactly free of modifications anyway, and an important downside is : now it's necessary to generate a (potentially large) table with names stored on heap, bringing with it the corresponding responsibility to manage the memory, aka to free it properly. It also makes interfaces more complex, with lifetime issues to document.

This seems more complex and, given a long-term maintenance perspective, more dangerous than the other solution, which concentrates the topic in a single place, thus largely freeing the rest of the system from having to deal with it.

Reply:
Yep that was definitely my initial approach - minimize my code footprint overall. However, I also see the merits in modifying FIO_compressMultipleFilenames() some more in favor of introducing a new `dstFilenameTable`. 

I think the approach then is to modify my function that takes a table and generates a table into a function that takes a `srcFileName` and `outDirName`, returns the final path, and call that in FIO_compressMultipleFilenames() at some point to modify our final file destination.

Reply:
Do you even need a `table` at all ?

Reply:
Nope! Not using new table at all in the new implementation, we only make a table of pointers when we need to check whether or not it there are name collisions (our approach is sorting the table and then checking for consecutive matches).